examples/stabilizer_code.py

```diff
@@ -0,0 +1,340 @@
+from typing import List, Tuple
+
+import numpy as np
+
+import cirq
+
+# Based on:
+# Stabilizer Codes and Quantum Error Correction
+# Daniel Gottesman
+# https://thesis.library.caltech.edu/2900/2/THESIS.pdf
+#
+# However, there were some mistakes in the thesis:
+# https://www2.perimeterinstitute.ca/personal/dgottesman/thesis-errata.html
+
+
+def _BuildByCode(mat: np.ndarray) -> List[str]:
+    """
+    Takes into input a matrix of Boolean interpreted as row-vectors, each having dimension 2 * n.
+    The matrix is converted into another matrix with as many rows, but this time the vectors
+    contain the letters I, X, Y, and Z representing Pauli operators.
+    """
+    out = []
+    n = mat.shape[1] // 2
+    for i in range(mat.shape[0]):
+        ps = ''
+        for j in range(n):
+            if mat[i, j] == 0 and mat[i, j + n] == 0:
+                ps += 'I'
+            elif mat[i, j] == 1 and mat[i, j + n] == 0:
+                ps += 'X'
+            elif mat[i, j] == 0 and mat[i, j + n] == 1:
+                ps += 'Z'
+            else:
+                ps += 'Y'
+        out.append(ps)
+    return out
+
+
+# It was considered to use scipy.linalg.lu but it seems to be only for real numbers and does
+# not allow to restrict only on a section of the matrix.
+def _GaussianElimination(
+    M: np.ndarray, min_row: int, max_row: int, min_col: int, max_col: int
+) -> int:
+    """
+    Performs a Gaussian elemination of the input matrix and transforms it into its reduced row
+    echelon form. The elimination is done only on a sub-section of the matrix (specified) by
+    ranges of rows and columns. The matrix elements are integers {0, 1} interpreted as elements
+    of GF(2).
+
+    In short, this is the implementation of section 4.1 of the thesis.
+
+    Args:
+        M: The input/output matrix
+        min_row: The minimum row (inclusive) where the perform the elimination.
+        max_row: The maximum row (exclusive) where the perform the elimination.
+        min_col: The minimum column (inclusive) where the perform the elimination.
+        max_col: The maximum column (exclusive) where the perform the elimination.
+
+    Returns:
+        The rank of the matrix.
+    """
+    assert M.shape[1] % 2 == 0
```

Comment:
Assert should be used for conditions that are believed impossible to arise. To signal invalid arguments, prefer to raise `ValueError`.

Reply:
It was my understanding of the difference between assert and ValueError indeed.

But that's what I intended to do, since I created M with: M = np.zeros((n - k, 2 * n), np.int8) and so it really should always have an even number of columns.